### ISSUE TITLE
Use IpEndpoint's FromStr implementation

### ISSUE THREAD

**Issue by dlrobertson** [redox-os/netstack#10]
m-labs/smoltcp#102 implmented [`FromStr`] for [`IpEndpoint`]. Currently [`parse_endpoint`] parses the port and IP address manually. We should use the [`FromStr`] implementation instead.

Note: we need to maintain the same defaults. AFAIK the [current upstream implementation] of the parser will return an error for `:80` while [`parse_endpoint`] will not.

[`FromStr`]: https://doc.rust-lang.org/std/str/trait.FromStr.html
[`IpEndpoint`]: https://docs.rs/smoltcp/0.4.0/smoltcp/wire/struct.IpEndpoint.html
[`parse_endpoint`]: https://github.com/redox-os/netstack/blob/master/src/smolnetd/scheme/mod.rs#L242-L255
[current upstream implementation]: https://github.com/m-labs/smoltcp/blob/master/src/parsers.rs#L330-L344

**Comment by batonius** [comment]
Indeed, currently `parse_endoint` is uber-permissive and accepts any string, basically using `0.0.0.0:0` as an error value. 

> Note: we need to maintain the same defaults.

I don't think we should bother with it, there's no reason for `smolnetd` to accept paths like ":80" or "abcd:80" or even "totally not an endpoint" (which would be parsed as `0.0.0.0:0`, prompting `smolnetd` to auto-assign a porn number and `smolnetd` to provide a valid src IP), I think the `parse_endopint` function should be removed and all the calls should be replaced with `FromStr::from_string`, early returning `syscall:EINVAL` on parsing errors. The only exception is empty strings, which should be treated as `0.0.0.0:0`.

**Comment by arellanj** [comment]
If its fine with you guys, I would like to work on his as my first issue.

**Comment by dlrobertson** [comment]
@arellanj awesome!
